mne/utils.py

```diff
@@ -583,6 +605,103 @@ def _compute_row_norms(data):
     return norms
 
 
+def reg_pinv(x, reg=0, rank=None, rcond=1e-15):
+    """Compute a regularized pseudoinverse of a square matrix.
```

Comment:
Is it right that this code not only moved but also changed a lot? If so, can you summarize the most important changes quickly? The ``diff`` is not of much help here ;-)

Reply:
In the beamformer code in the current master branch, we use three methods to invert a matrix: `_reg_pinv`, `_eig_inv` and `scipy.linalg.pinv`. These three functions actually do something very similar: they invert a matrix through SVD- (or Eigen-) decomposition and only using *n* components in the inverse (=rank). Even `scipy.linalg.pinv` does nothing more than that. Where the functions differ is in the method of determining how many components to use.

So I thought it best to refactor `_reg_pinv` to a function that makes this explicit. It offers precise control over the number of components to use and thereby fulfils all the needs we have without having different inverse functions all over the place.

Importantly it offers these ways to set the rank (=number of components to use in inverse):

 1. Set it explicitly
 2. Use a threshold for the singular values *before* adding the regularization
 3. Use a threshold for the singular values *after* adding the regularization

Reply:
To check my understanding, (3) will produce a full rank output for any reg value not closer to zero than the threshold (i.e., not tiny), right?

Reply:
Cool. I'd be interested to hear if you have any thoughts about https://github.com/mne-tools/mne-python/pull/5481 since it seems closely related

Reply:
Aha, my bad. I think I got the notion of changes from ``reg_pinv`` being listed in New Features above ;-)